base/glibc-compatibility/musl/getauxval.c

```diff
@@ -1,59 +1,74 @@
 #include <sys/auxv.h>
+#include <fcntl.h> // open
+#include <sys/stat.h> // O_RDONLY
+#include <unistd.h> // read, close
+#include <stdlib.h> // ssize_t
+#include <stdio.h> // perror, fprintf
+#include <link.h> // ElfW
 #include "atomic.h"
-#include <unistd.h> // __environ
 #include <errno.h>
 
-// We don't have libc struct available here. Compute aux vector manually.
-static unsigned long * __auxv = NULL;
-static unsigned long __auxv_secure = 0;
+#define ARRAY_SIZE(a) sizeof((a))/sizeof((a[0]))
 
-static size_t __find_auxv(unsigned long type)
-{
-    size_t i;
-    for (i = 0; __auxv[i]; i += 2)
-    {
-        if (__auxv[i] == type)
-            return i + 1;
-    }
-    return (size_t) -1;
-}
+// We don't have libc struct available here.
+// Compute aux vector manually (from /proc/self/auxv).
+static unsigned long __auxv[64];
+static unsigned long __auxv_secure = 0;
 
 unsigned long __getauxval(unsigned long type)
 {
     if (type == AT_SECURE)
         return __auxv_secure;
 
-    if (__auxv)
+    if (type >= ARRAY_SIZE(__auxv))
     {
-        size_t index = __find_auxv(type);
-        if (index != ((size_t) -1))
-            return __auxv[index];
+        errno = ENOENT;
+        return 0;
     }
 
-    errno = ENOENT;
-    return 0;
+    return __auxv[type];
 }
 
 static void * volatile getauxval_func;
 
-static unsigned long  __auxv_init(unsigned long type)
+ssize_t __retry_read(int fd, void *buf, size_t count)
 {
-    if (!__environ)
+    for (;;)
     {
-        // __environ is not initialized yet so we can't initialize __auxv right now.
-        // That's normally occurred only when getauxval() is called from some sanitizer's internal code.
-        errno = ENOENT;
-        return 0;
+        int ret = read(fd, buf, count);
+        if (ret == -1)
+        {
+            if (errno == EINTR)
+                continue;
+            perror("Cannot read /proc/self/auxv");
+            abort();
+        }
+        return ret;
+    }
+}
+static unsigned long  __auxv_init(unsigned long type)
+{
+    // od -t dL /proc/self/auxv
+    int fd = open("/proc/self/auxv", O_RDONLY);
+    if (fd == -1) {
+        perror("Cannot read /proc/self/auxv (likely kernel is too old or procfs is not mounted)");
+        abort();
     }
 
-    // Initialize __auxv and __auxv_secure.
-    size_t i;
-    for (i = 0; __environ[i]; i++);
-    __auxv = (unsigned long *) (__environ + i + 1);
+    ElfW(auxv_t) aux;
+    while (__retry_read(fd, &aux, sizeof(aux)) == sizeof(aux))
```

Comment:
Ok but we need a comment or a check or correct handling,
that we don't expect `read` to read less than `sizeof(aux)`.

Reply:
Added a comment.